From 17179c66c8d1f5931a8700be5ee17e07a58757c0 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Wed, 20 Dec 2023 10:48:18 +0800
Subject: [PATCH 10/94] renderer-gl: Fix segfault when reading pixels

We should use the actual stride to create tmp pixmap.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 libweston/renderer-gl/gl-renderer.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/libweston/renderer-gl/gl-renderer.c b/libweston/renderer-gl/gl-renderer.c
index 2d98cc7d05..c1d4fef41e 100644
--- a/libweston/renderer-gl/gl-renderer.c
+++ b/libweston/renderer-gl/gl-renderer.c
@@ -796,6 +796,7 @@ gl_renderer_do_read_pixels(struct gl_renderer *gr,
 {
 	void *read_target;
 	pixman_image_t *tmp = NULL;
+	void *tmp_data = NULL;
 
 	assert(fmt->gl_type != 0);
 	assert(fmt->gl_format != 0);
@@ -811,11 +812,17 @@ gl_renderer_do_read_pixels(struct gl_renderer *gr,
 		 * glReadPixels() returns bottom row first. We need to
 		 * read into a temporary buffer and y-flip it.
 		 */
+		tmp_data = malloc(stride * rect->height);
+		if (!tmp_data)
+			return false;
+
 		tmp = pixman_image_create_bits(fmt->pixman_format,
 					       rect->width, rect->height,
-					       NULL, 0);
-		if (!tmp)
+					       tmp_data, stride);
+		if (!tmp) {
+			free(tmp_data);
 			return false;
+		}
 
 		read_target = pixman_image_get_data(tmp);
 	}
@@ -850,6 +857,7 @@ gl_renderer_do_read_pixels(struct gl_renderer *gr,
 
 		pixman_image_unref(image);
 		pixman_image_unref(tmp);
+		free(tmp_data);
 	}
 
 	return true;
-- 
2.20.1

